### PR TITLE
feat(RHTAPWATCH-580): Add deployment and service resources

### PR DESCRIPTION
**Pull request by hmariset**
Add deployment and service kubernetes resources to deploy
the dummy exporter
    
    Signed-off-by: Homaja Marisetty <hmariset@redhat.com>

### CODE REVIEW

File: config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      app: prometheus-exporter
+  template:
+    metadata:
+      labels:
+        app: prometheus-exporter
+    spec:
+      serviceAccountName: exporter-sa
+      containers:
+      - name: prometheus-exporter
+        image: quay.io/redhat-appstudio / o11y-prometheus-exporters:latest
```
Comment (avi-biton) [inline config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml:60]:
```suggestion
        image: quay.io/redhat-appstudio/o11y-prometheus-exporters:latest
```

**Comment by avi-biton** [comment]
I am failing to pull this image in development environment, I get this error:
`Failed to pull image "quay.io/redhat-appstudio/o11y-prometheus-exporters:latest": rpc error: code = Unknown desc = reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized`

**Comment by hmariset** [comment]
> I am failing to pull this image in development environment, I get this error:
> 
> ```
> Failed to pull image "quay.io/redhat-appstudio/o11y-prometheus-exporters:latest": rpc error: code = Unknown desc = reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized
> ```

@avi-biton, When I pushed the image it created private repository. I think I need to add manually to give permissions. Is your quay userId abiton?

**Comment by avi-biton** [comment]
> > I am failing to pull this image in development environment, I get this error:
> > ```
> > Failed to pull image "quay.io/redhat-appstudio/o11y-prometheus-exporters:latest": rpc error: code = Unknown desc = reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized
> > ```
> 
> @avi-biton, When I pushed the image it created private repository. I think I need to add manually to give permissions. Is your quay userId abiton?

Its `abiton1`

File: config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+metadata:
+  name: exporter-cluster-role-binding
+subjects:
+- kind: ServiceAccount
+  name: exporter-sa
+  namespace: appstudio-exporter-service-prometheus
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: exporter-cluster-role
```
Comment (avi-biton) [inline config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml:27]:
```suggestion
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: exporter-cluster-role-binding
  namespace: appstudio-grafana
subjects:
- kind: ServiceAccount
  name: exporter-sa
  namespace: appstudio-exporter-service-prometheus
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: exporter-cluster-role
```

This way we restrict the access defined in the `ClusterRole` only to `appstudio-grafana` namespace

**Comment by hmariset** [comment]
> > > I am failing to pull this image in development environment, I get this error:
> > > ```
> > > Failed to pull image "quay.io/redhat-appstudio/o11y-prometheus-exporters:latest": rpc error: code = Unknown desc = reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized
> > > ```
> > 
> > 
> > @avi-biton, When I pushed the image it created private repository. I think I need to add manually to give permissions. Is your quay userId abiton?
> 
> Its `abiton1`

Added you..I have added o11y group in the permissions but I think you aren't there in that. Please check now.

File: config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml (hunk 1 above)
Comment (hmariset) [reply config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml]: Already pushed the change, but due to Github issues it took sometime to reflect the change.

File: config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml (hunk 2 above)
Comment (hmariset) [reply config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml:27]: Ok, will update it.

**Comment by yftacherzog** [comment]
> > I am failing to pull this image in development environment, I get this error:
> > ```
> > Failed to pull image "quay.io/redhat-appstudio/o11y-prometheus-exporters:latest": rpc error: code = Unknown desc = reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized
> > ```
> 
> @avi-biton, When I pushed the image it created private repository. I think I need to add manually to give permissions. Is your quay userId abiton?

The RHTAP clusters should be able to pull it. I'm not sure adding our group will cut it. We need to see what sort of permissions other images for RHTAP have and use something similar.

File: config/exporters/monitoring/prometheus/base/kustomization.yaml
Comment (yftacherzog) [inline config/exporters/monitoring/prometheus/base/kustomization.yaml:1]: I'm not sure the path makes sense to me. If we follow the logic for the path used for the probe then it should be `.../exporters/monitoring/grafana/...` as it monitors grafana stuff.

@avi-biton any suggestions (as you were the one to set the path for the probe)?

File: config/exporters/monitoring/prometheus/base/namespace.yaml
Comment (yftacherzog) [inline config/exporters/monitoring/prometheus/base/namespace.yaml:1]: What's the reason for having this on a separate file?

**Comment by yftacherzog** [comment]
> > I am failing to pull this image in development environment, I get this error:
> > ```
> > Failed to pull image "quay.io/redhat-appstudio/o11y-prometheus-exporters:latest": rpc error: code = Unknown desc = reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized
> > ```
> 
> @avi-biton, When I pushed the image it created private repository. I think I need to add manually to give permissions. Is your quay userId abiton?

@hmariset are you able to have it deployed on your dev env?

**Comment by avi-biton** [comment]
@yftacherzog Do we want the exporter to be in a separate namespace than the service it tests?
In other words, don't we want this exporter to be in the `appstudio-grafana` namespace?

File: config/exporters/monitoring/prometheus/base/kustomization.yaml
Comment (avi-biton) [reply config/exporters/monitoring/prometheus/base/kustomization.yaml:1]: I tend to agree with you on this one, following the same logic we can go with `../exporters/monitoring/grafana/...`
`prometheus` is irrelevant

**Comment by yftacherzog** [comment]
> @yftacherzog Do we want the exporter to be in a separate namespace than the service it tests? In other words, don't we want this exporter to be in the `appstudio-grafana` namespace?

I don't recall anymore what we had in mind. With the probes we definitely wanted  them on different namespaces. With the exporters approach, I'm not sure anymore. I tend to think that it's better to have them on different namespaces, but not sure if we had it clear.

@amisstea what do you say?

**Comment by avi-biton** [comment]
> > > I am failing to pull this image in development environment, I get this error:
> > > ```
> > > Failed to pull image "quay.io/redhat-appstudio/o11y-prometheus-exporters:latest": rpc error: code = Unknown desc = reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized
> > > ```
> > 
> > 
> > @avi-biton, When I pushed the image it created private repository. I think I need to add manually to give permissions. Is your quay userId abiton?
> 
> The RHTAP clusters should be able to pull it as well as the developers. I'm not sure adding our group will cut it. We need to see what sort of permissions other images for RHTAP have and use something similar.

That's right, it should not accessible to all developers, we should not add anything manually.

BTW, I am able to reach the image in quay https://quay.io/repository/redhat-appstudio/o11y-prometheus-exporters?tab=info
And still, I cannot pull the image in my development environment or by using `podman pull`, and I still get this message:
```
[abiton@abiton ~]$ podman pull quay.io/redhat-appstudio/o11y-prometheus-exporters
Trying to pull quay.io/redhat-appstudio/o11y-prometheus-exporters:latest...
Error: initializing source docker://quay.io/redhat-appstudio/o11y-prometheus-exporters:latest: reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized
```

**Comment by hmariset** [comment]
> podman pull quay.io/redhat-appstudio/o11y-prometheus-exporters



> 

@avi-biton if I didn't login to quay.io in terminal, I have this error. Once I logged in, I was able to pull the image.

**Comment by hmariset** [comment]
> > > I am failing to pull this image in development environment, I get this error:
> > > ```
> > > Failed to pull image "quay.io/redhat-appstudio/o11y-prometheus-exporters:latest": rpc error: code = Unknown desc = reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized
> > > ```
> > 
> > 
> > @avi-biton, When I pushed the image it created private repository. I think I need to add manually to give permissions. Is your quay userId abiton?
> 
> @hmariset are you able to have it deployed on your dev env?

@yftacherzog, @avi-biton  I have deployed the `prometheus-exporter-service.yaml` file in dev env but the pod is failing with `CrashLoopBackOff` error. I see the image was pulled successfully but failing to restart the container.

**Comment by yftacherzog** [comment]
> > > > I am failing to pull this image in development environment, I get this error:
> > > > ```
> > > > Failed to pull image "quay.io/redhat-appstudio/o11y-prometheus-exporters:latest": rpc error: code = Unknown desc = reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized
> > > > ```
> > > 
> > > 
> > > @avi-biton, When I pushed the image it created private repository. I think I need to add manually to give permissions. Is your quay userId abiton?
> > 
> > @hmariset are you able to have it deployed on your dev env?
> 
> @yftacherzog, @avi-biton  I have deployed the `prometheus-exporter-service.yaml` file in dev env but the pod is failing with `CrashLoopBackOff` error. I see the image was pulled successfully but failing to restart the container.  

The pod events tab should be able to provide some details for the failure.

**Comment by hmariset** [comment]
> > > > > I am failing to pull this image in development environment, I get this error:
> > > > > ```
> > > > > Failed to pull image "quay.io/redhat-appstudio/o11y-prometheus-exporters:latest": rpc error: code = Unknown desc = reading manifest latest in quay.io/redhat-appstudio/o11y-prometheus-exporters: unauthorized: access to the requested resource is not authorized
> > > > > ```
> > > > 
> > > > 
> > > > @avi-biton, When I pushed the image it created private repository. I think I need to add manually to give permissions. Is your quay userId abiton?
> > > 
> > > 
> > > @hmariset are you able to have it deployed on your dev env?
> > 
> > 
> > @yftacherzog, @avi-biton  I have deployed the `prometheus-exporter-service.yaml` file in dev env but the pod is failing with `CrashLoopBackOff` error. I see the image was pulled successfully but failing to restart the container.
> 
> The pod events tab should be able to provide some details for the failure.

In the events tab only I see pulled image successfully then `Back-off restarting failed container`.

File: config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml [hunk 3]
```diff
@@ -0,0 +1,74 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: exporter-sa
+  namespace: appstudio-exporter-service-prometheus
```
Comment (avi-biton) [inline config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml:5]: I would suggest `appstudio-grafana-datasource-exporter`

File: config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+metadata:
+  name: exporter-service
+  namespace: appstudio-exporter-service-prometheus
+spec:
+  ports:
+  - name: https
+    port: 8090
+    targetPort: https
+  selector:
+    app: prometheus-exporter
```
Comment (avi-biton) [inline config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml:41]:
```suggestion
    app: grafana-datasource-exporter
```

File: config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: exporter-service-deployment
+  namespace: appstudio-exporter-service-prometheus
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: prometheus-exporter
```
Comment (avi-biton) [inline config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml:52]:
```suggestion
      app: grafana-datasource-exporter
```

File: config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  template:
+    metadata:
+      labels:
+        app: prometheus-exporter
+    spec:
+      serviceAccountName: exporter-sa
+      containers:
+      - name: prometheus-exporter
```
Comment (avi-biton) [inline config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml:60]:
```suggestion
        app: grafana-datasource-exporter
```

**Comment by hmariset** [comment]
> podman run quay.io/redhat-appstudio/o11y-prometheus-exporters:latest

@avi-biton I tried running other images from redhat-appstudio, they also returned immediately to prompt
`[hmariset@hmariset-thinkpadp1gen3 o11y]$ podman run quay.io/redhat-appstudio/buildah:v1.28
[hmariset@hmariset-thinkpadp1gen3 o11y]$ `

**Comment by yftacherzog** [comment]
@hmariset the main thing that is broken here is the Dockerfile. The structure that we have there is:

```
FROM registry.access.redhat.com/ubi9/go-toolset:1.20.10-2 AS builder

<here we create the stuff that we expect to be on the image>
...
FROM registry.access.redhat.com/ubi8/ubi-minimal:8.6-751
RUN microdnf update --setopt=install_weak_deps=0 -y && microdnf install libcurl-minimal libcurl-devel
```

The last two lines are taking a new image, installing some dnf packages on top of it and publishing it.
It's not taking anything from the builder image, so we're actually publishing the ubi8 image with some dnf packages installed on it, while the part of the Dockerfile in which we build our binaries is not being used.

Can you create a bug in Jira to fix our image?

If we need those packages from the last two lines installed/updated then we need both them and our binaries on the image that we publish (the image appearing last on the Dockerfile).

**Comment by hmariset** [comment]
@amisstea @avi-biton @yftacherzog I have updated the Dockerfile and now the pod is running and in available status. Please review again.

File: config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml (hunk 3 above)
Comment (hmariset) [reply config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml]: Updated to `appstudio-grafana-datasource-exporter`

File: config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml (hunk 4 above)
Comment (hmariset) [reply config/exporters/monitoring/prometheus/base/prometheus-exporter-service.yaml]: Updated to `grafana-datasource-exporter`

**Comment by yftacherzog** [comment]
> @amisstea @avi-biton @yftacherzog I have updated the Dockerfile and now the pod is running and in available status. Please review again.

Can you explain the test you did? Where did you test and with what image?

**Comment by hmariset** [comment]
> > @amisstea @avi-biton @yftacherzog I have updated the Dockerfile and now the pod is running and in available status. Please review again.
> 
> Can you explain the test you did? Where did you test and with what image?

I have updated the image with modified Dockerfile and pushed it to quay repository. Tested locally with `podman run -p 8090:8090 quay.io/redhat-appstudio/o11y-prometheus-exporters:latest` command and it is working. Also, re-deployed  `prometheus-exporter-service.yaml `  file on dev environment and checked the pod status which is in running status.

**Comment by yftacherzog** [comment]
> > > @amisstea @avi-biton @yftacherzog I have updated the Dockerfile and now the pod is running and in available status. Please review again.
> > 
> > Can you explain the test you did? Where did you test and with what image?
> 
> I have updated the image with modified Dockerfile and pushed it to quay repository. Tested locally with `podman run -p 8090:8090 quay.io/redhat-appstudio/o11y-prometheus-exporters:latest` command and it is working. Also, re-deployed  `prometheus-exporter-service.yaml `  file on dev environment and checked the pod status which is in running status.

Cool. I suggest also configure port forwarding for the service and verify that you're able to reach the metrics endpoint

**Review by yftacherzog — DISMISSED** [review]

File: config/exporters/monitoring/prometheus/base/kustomization.yaml
Comment (hmariset) [reply config/exporters/monitoring/prometheus/base/kustomization.yaml:1]: Updated it.

**Review by avi-biton — APPROVED** [review]
LGTM

File: Dockerfile [hunk 7]
```diff
@@ -9,11 +9,13 @@ COPY go.mod go.sum ./
...
 COPY --chown=default:root exporters/**/*.go .
 
-RUN ["go", "build", "."]
+RUN CGO_ENABLED=0 GOOS=linux go build -o /bin/exporters .
 
 EXPOSE 8090
-CMD ["go", "run", "."]
 
 
 FROM registry.access.redhat.com/ubi8/ubi-minimal:8.6-751
```
Comment (yftacherzog) [inline Dockerfile:17]: I wonder what's the reasoning behind using ubi9-based image for the builder and ubi8-based for the final image. Can we use ubi9 for both?

Comment (hmariset) [reply Dockerfile:17]: I tried using ubi9/ubi-minimal:9.2-750.1697625013 but it was stuck for a long time after installing some packages, so I used ubi8 for the second image. For the first image, Go Toolset for ubi9 contains the fix for CVE-2023-44487 so I have used ubi9 for first one.

Comment (yftacherzog) [reply Dockerfile:17]: I think it's stuck because it's missing the `-y` on the install command.
It's waiting for user confirmation.

**Review by yftacherzog — APPROVED** [review]

**Pull request merged**